File: Python/examples/cashflows.py
# ---
# jupyter:
#   jupytext:
#     formats: py:percent
#     text_representation:
#       extension: .py
#       format_name: percent
#       format_version: '1.3'
#       jupytext_version: 1.6.0
#   kernelspec:
#     display_name: Python 3
#     language: python
#     name: python3
# ---

# %% [markdown]
# # Cash-flow analysis
#
# Copyright (&copy;) 2020 StatPro Italia srl
#
# This file is part of QuantLib, a free-software/open-source library
# for financial quantitative analysts and developers - https://www.quantlib.org/
#
# QuantLib is free software: you can redistribute it and/or modify it under the
# terms of the QuantLib license.  You should have received a copy of the
# license along with this program; if not, please email
# <quantlib-dev@lists.sf.net>. The license is also available online at
# <https://www.quantlib.org/license.shtml>.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the license for more details.

# %%
import QuantLib as ql
import pandas as pd

# %%
interactive = "get_ipython" in globals()

# %%
today = ql.Date(19, ql.October, 2020)
ql.Settings.instance().evaluationDate = today

# %% [markdown]
# ### Term structure construction

# %%
dates = [
    ql.Date(19,10,2020),
    ql.Date(19,11,2020),
    ql.Date(19, 1,2021),
    ql.Date(19, 4,2021),
    ql.Date(19,10,2021),
    ql.Date(19, 4,2022),
    ql.Date(19,10,2022),
    ql.Date(19,10,2023),
    ql.Date(19,10,2025),
    ql.Date(19,10,2030),
    ql.Date(19,10,2035),
    ql.Date(19,10,2040),
]

rates = [
    -0.004,
    -0.002,
    0.001,
    0.005,
    0.009,
    0.010,
    0.010,
    0.012,
    0.017,
    0.019,
    0.028,
    0.032,
]

forecast_curve = ql.ZeroCurve(dates, rates, ql.Actual365Fixed())

# %%
forecast_handle = ql.YieldTermStructureHandle(forecast_curve)

# %% [markdown]
# ### Swap construction
#
# We'll use an overnight swap as an example.  We're keeping the initialization simple, but the analysis work in the same way for more complex ones, as well as for other kinds of swaps and bonds (once we extract the cashflows from them using the proper methods).

# %%
swap = ql.MakeOIS(swapTenor=ql.Period(5, ql.Years),
                  overnightIndex=ql.Eonia(forecast_handle),
                  fixedRate=0.002)

# %% [markdown]
# ### Cash-flow analysis
#
# The fixed-rate coupons can be extracted from the swap using the `fixedLeg` method.  They are returned as instances of the base `Cashflow` class, so the only methods we have directly available are from that class interface:

# %%
fixed_leg = swap.fixedLeg()

# %%
df = pd.DataFrame([(c.date(), c.amount()) for c in fixed_leg if c.date() > today],
                  columns=['date', 'amount'])
df

# %% [markdown]
# The following displays the results when this is run as a Python script (in which case the cell above is not displayed).

# %%
if not interactive:
    print(df)

# %% [markdown]
# If we want to extract more information, we need to upcast the coupons to a more specific class.  This can be done by using the `as_fixed_rate_coupon` method.  In this case, the upcast works by construction; but in the general case we might have cashflows for which the upcast fails (e.g., the redemption for a bond) so we have to check for nulls.

# %%
coupons = []
for cf in fixed_leg:
    c = ql.as_fixed_rate_coupon(cf)
    if c:
        coupons.append(c)

# %% [markdown]
# We can now access methods from the coupon class.

# %%
df = pd.DataFrame([(c.date(), c.amount(), c.rate(), c.accrualStartDate(), c.accrualEndDate(), c.accrualPeriod())
                   for c in coupons if c.date() > today],
                  columns=['payment date', 'amount', 'rate', 'start date', 'end date', 'accrual period'])
df

# %%
if not interactive:
    print(df)

# %% [markdown]
# The same goes for the floating leg: in this case, we need to upcast to floating-rate coupons in order to access the specific methods we'll need.

# %%
floating_leg = swap.overnightLeg()

# %%
coupons = []
for cf in floating_leg:
    c = ql.as_floating_rate_coupon(cf)
    if c:
        coupons.append(c)

# %%
df = pd.DataFrame([(c.date(), c.amount(), c.rate(), c.accrualStartDate(), c.accrualEndDate(), c.accrualPeriod())
                   for c in coupons if c.date() > today],
                  columns=['payment date', 'amount', 'rate', 'start date', 'end date', 'accrual period'])
df

# %%
if not interactive:
    print(df)

File: R/demo/cashflows.R

# inspired by python example with the same name

suppressMessages(library(QuantLib))

## global data
calendar <- TARGET()

todaysDate <- Date(19, "October", 2020)
invisible(Settings_instance()$setEvaluationDate(d=todaysDate))

settlementDays <- 3
settlementDate <- Calendar_advance(calendar, todaysDate, settlementDays, "Days")

cat('Today          : ', todaysDate$`__str__`(), "\n")
cat('Settlement Date: ', settlementDate$`__str__`(), "\n")

# ### Term structure construction

# %%
dates = DateVector()
DateVector_append(dates, DateParser_parseISO("2020-10-19"))
DateVector_append(dates, DateParser_parseISO("2020-11-19"))

DateVector_append(dates, DateParser_parseISO("2021-01-19"))
DateVector_append(dates, DateParser_parseISO("2021-04-19"))
DateVector_append(dates, DateParser_parseISO("2021-10-19"))

DateVector_append(dates, DateParser_parseISO("2022-04-19"))
DateVector_append(dates, DateParser_parseISO("2022-10-19"))

DateVector_append(dates, DateParser_parseISO("2023-10-19"))
DateVector_append(dates, DateParser_parseISO("2025-10-19"))
DateVector_append(dates, DateParser_parseISO("2030-10-19"))
DateVector_append(dates, DateParser_parseISO("2035-10-19"))
DateVector_append(dates, DateParser_parseISO("2040-10-19"))

DateVector_size(dates)

rates = c(
    -0.004,
    -0.002,
    0.001,
    0.005,
    0.009,
    0.010,
    0.010,
    0.012,
    0.017,
    0.019,
    0.028,
    0.032
)

length(rates)

forecast_curve = ZeroCurve(dates, rates, Actual365Fixed())
forecast_handle = YieldTermStructureHandle(forecast_curve)

# ### Swap construction
#
# We'll use an overnight swap as an example.  
# We're keeping the initialization simple, 
# but the analysis work in the same way for more complex ones, 
# as well as for other kinds of swaps and bonds (once we extract the cashflows from them using the proper methods).

# %%
swapBuilder = MakeOIS(swapTenor=Period(5, "Years"),
                  overnightIndex=Eonia(forecast_handle),
                  fixedRate=0.002)

swap = MakeOIS_makeOIS(swapBuilder)

# ### Cash-flow analysis
#
# The fixed-rate coupons can be extracted from the swap using the `fixedLeg` method.  
#

fixed_leg = swap$fixedLeg()
CashFlows_maturityDate(fixed_leg)

df = NULL
for (i in seq(1, fixed_leg$size())) {
  cfl = fixed_leg[i][[1]]
  df = rbind(df, data.frame(date=Date_ISO(CashFlow_date(cfl)), amount=CashFlow_amount(cfl)))
}
print(df)


#
# If we want to extract more information, we need to upcast the coupons to a more specific class.  
# This can be done by using the `as_coupon` or the `as_fixed_rate_coupon` method. 
#

df = NULL
for (i in seq(1, fixed_leg$size())) {
  cfl = fixed_leg[i][[1]]
  cflCoupon = as_coupon(cfl)
  cflCouponFix = as_fixed_rate_coupon(cfl)
  cflCouponFixIr = FixedRateCoupon_interestRate(cflCouponFix)
  
  df = rbind(df, data.frame(date=Date_ISO(CashFlow_date(cfl)), amount=CashFlow_amount(cfl), 
                            accrualStartDate=Date_ISO(Coupon_accrualStartDate(cflCoupon)),
                            accrualEndDate=Date_ISO(Coupon_accrualEndDate(cflCoupon)),
                            accrualPeriod=Coupon_accrualPeriod(cflCoupon),
                            accrualDays=Coupon_accrualDays(cflCoupon),
                            accrualDayCounter=DayCounter_name(Coupon_dayCounter(cflCoupon)),
                            accruedAmount=Coupon_accruedAmount(self = cflCoupon, todaysDate),
                            rate=InterestRate_rate(cflCouponFixIr), rateDayCount=DayCounter_name(InterestRate_dayCounter(cflCouponFixIr))
                            ))
}
print(df)



# %%
floating_leg = swap$overnightLeg()

# %%
df = NULL
for (i in seq(1, floating_leg$size())) {
  cfl = floating_leg[i][[1]]
  cflCoupon = as_coupon(cfl)
  cflCouponFloat = as_floating_rate_coupon(cfl)
  cflCouponFloatIdx = FloatingRateCoupon_index(cflCouponFloat)
  
  
  df = rbind(df, data.frame(date=Date_ISO(CashFlow_date(cfl)), amount=CashFlow_amount(cfl), 
                            accrualStartDate=Date_ISO(Coupon_accrualStartDate(cflCoupon)),
                            accrualEndDate=Date_ISO(Coupon_accrualEndDate(cflCoupon)),
                            accrualPeriod=Coupon_accrualPeriod(cflCoupon),
                            accrualDays=Coupon_accrualDays(cflCoupon),
                            accrualDayCounter=DayCounter_name(Coupon_dayCounter(cflCoupon)),
                            accruedAmount=Coupon_accruedAmount(self = cflCoupon, todaysDate),
                            spread=FloatingRateCoupon_spread(cflCouponFloat),
                            gearing=FloatingRateCoupon_gearing(cflCouponFloat),
                            adjFix=FloatingRateCoupon_adjustedFixing(cflCouponFloat),
                            dtFix=Date_ISO(FloatingRateCoupon_fixingDate(cflCouponFloat)),
                            convexityAdj=FloatingRateCoupon_convexityAdjustment(cflCouponFloat)
  ))
}

print(df)

